3p-ecosystem/src/main/java/com/google/homesampleapp/screens/settings/SettingsDeveloperUtilitiesScreen.kt

```diff
@@ -0,0 +1,187 @@
+  // When the user dismisses the Msg AlertDialog, we "consume" the dialog.
+  val msgDialogInfo by developerUtilitiesViewModel.msgDialogInfo.collectAsState()
+  val onDismissMsgDialog: () -> Unit = { developerUtilitiesViewModel.dismissMsgDialog() }
+
+  // Log Repos
+  val showLogReposDialog by developerUtilitiesViewModel.showLogReposDialog.collectAsState()
+  val onShowLogReposDialog: () -> Unit = { developerUtilitiesViewModel.printRepositories() }
+  val onDismissLogReposDialog: () -> Unit = {
+    developerUtilitiesViewModel.dismissLogRepositoriesDialog()
+  }
+
+  // Showing the commissionable devices requires scanning permissions.
+  // This defines a launcher for the activity that requests the user to
+  // allow the required permissions.
+  val scanningPermissionsLauncher =
+    rememberLauncherForActivityResult(
+      contract = ActivityResultContracts.RequestMultiplePermissions()
+    ) { results ->
+      // Process the "RequestMultiplePermissions" activity results
+      if (results.values.any { granted -> !granted }) {
+        // At least one permission is not granted.
+        Timber.d("scanningPermissionsLauncher: All scanning permissions needed were not granted.")
+        developerUtilitiesViewModel.showMsgDialog(
+          "Scanning Permissions",
+          "Scanning permissions were not granted, so unfortunately " +
+            "the \"Commissionable Devices\" feature is not available.",
+        )
+      } else {
+        Timber.d("scanningPermissionsLauncher: Permissions were OK!")
+        navController.navigate("commissionable_devices")
+      }
+    }
+
+  val onCommissionableDevicesClick: () -> Unit = {
+    developerUtilitiesViewModel.logScanningPermissions(activity!!.applicationContext)
+    if (!developerUtilitiesViewModel.allScanningPermissionsGranted(activity.applicationContext)) {
+      Timber.d("All scanning permissions NOT granted. Asking for them.")
+      scanningPermissionsLauncher.launch(
+        developerUtilitiesViewModel.getRequiredScanningPermissions()
+      )
+    } else {
+      // Check if Bluetooth is enabled.
+      val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
+      if (bluetoothAdapter.isEnabled) {
+        navController.navigate("commissionable_devices")
+      } else {
+        Timber.d("Bluetooth is not enabled") // FIXME
+        developerUtilitiesViewModel.showMsgDialog(
+          "Bluetooth is not enabled",
+          "Bluetooth must be enabled on your phone to allow discovery of matter devices",
+        )
+      }
+    }
+  }
+
+  val onThreadClick: () -> Unit = { navController.navigate("thread") }
+
+  SettingsDeveloperUtilitiesScreen(
+    navController,
+    innerPadding,
+    msgDialogInfo,
+    onDismissMsgDialog,
+    showLogReposDialog,
+    onShowLogReposDialog,
+    onDismissLogReposDialog,
+    onCommissionableDevicesClick,
+    onThreadClick,
+  )
+}
+
+@Composable
+private fun SettingsDeveloperUtilitiesScreen(
+  navController: NavController,
```

Comment:
The NavController is an unstable type, so some indirection like a lambda should be used as the compiler considers lambdas stable. It should also help with previews.